Sources/Commands/ToolWorkspaceDelegate.swift

```diff
@@ -135,12 +135,16 @@ class ToolWorkspaceDelegate: WorkspaceDelegate {
         self.outputHandler("Computed \(location) at \(version) (\(duration.descriptionInSeconds))", false)
     }
 
-    func willDownloadBinaryArtifact(from url: String) {
-        self.outputHandler("Downloading binary artifact \(url)", false)
+    func willDownloadBinaryArtifact(from url: String, fromCache: Bool) {
+        if fromCache {
+            self.outputHandler("Fetching binary artifact \(url) from cache", false)
+        } else {
+            self.outputHandler("Downloading binary artifact \(url)", false)
+        }
     }
 
-    func didDownloadBinaryArtifact(from url: String, result: Result<AbsolutePath, Error>, duration: DispatchTimeInterval) {
-        guard case .success = result, !self.observabilityScope.errorsReported else {
+    func didDownloadBinaryArtifact(from url: String, result: Result<(path: AbsolutePath, fromCache: Bool), Error>, duration: DispatchTimeInterval) {
```

Comment:
I'm unsure about the use of `Result` here in the first place. Why would `didDownloadBinaryArtifact` called in the `failure` case of `Result`? Wouldn't it be better to have a separate `didFailBinaryArtifactDownload` method for that? And then the use of tuple here exacerbates the problem. If it were two separate delegate methods you wouldn't need a tuple.

Reply:
mostly I was trying to keep the set of changes more minimal, but I dont see huge value in a result here too. do you think we should expand the scope of the PR, or a separate one?

Reply:
I've been meaning to starting converting delegates to `AsyncSequence` APIs as a part of our concurrency changes. Then probably not worth touching it here or in a separate PR if it's going to be fixed in an imminent overhaul